components/suggestion/search_channel_with_permissions_provider.jsx

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {
+    getChannelsInCurrentTeam,
+} from 'mattermost-redux/selectors/entities/channels';
+import {getMyChannelMemberships} from 'mattermost-redux/selectors/entities/common';
+import {getConfig} from 'mattermost-redux/selectors/entities/general';
+import {getCurrentTeamId} from 'mattermost-redux/selectors/entities/teams';
+import * as ChannelActions from 'mattermost-redux/actions/channels';
+import {getCurrentUserLocale} from 'mattermost-redux/selectors/entities/i18n';
+import {haveICurrentTeamPermission} from 'mattermost-redux/selectors/entities/roles';
+import {Permissions} from 'mattermost-redux/constants';
+import {sortChannelsByTypeAndDisplayName} from 'mattermost-redux/utils/channel_utils';
+
+import GlobeIcon from 'components/svg/globe_icon';
+import LockIcon from 'components/svg/lock_icon';
+import ArchiveIcon from 'components/svg/archive_icon';
+import AppDispatcher from 'dispatcher/app_dispatcher.jsx';
+import store from 'stores/redux_store.jsx';
+import {ActionTypes, Constants} from 'utils/constants.jsx';
+
+import Provider from './provider.jsx';
+import Suggestion from './suggestion.jsx';
+
+class SearchChannelWithPermissionsSuggestion extends Suggestion {
+    static get propTypes() {
+        return {
+            ...super.propTypes,
+        };
+    }
+
+    render() {
+        const {item, isSelection} = this.props;
+        const channel = item.channel;
+        const channelIsArchived = channel.delete_at && channel.delete_at !== 0;
+
+        let className = 'mentions__name';
+        if (isSelection) {
+            className += ' suggestion--selected';
+        }
+
+        const displayName = channel.display_name;
+        let icon = null;
+        if (channelIsArchived) {
+            icon = (
+                <ArchiveIcon className='icon icon__archive'/>
+            );
+        } else if (channel.type === Constants.OPEN_CHANNEL) {
+            icon = (
+                <GlobeIcon className='icon icon__globe icon--body'/>
+            );
+        } else if (channel.type === Constants.PRIVATE_CHANNEL) {
+            icon = (
+                <LockIcon className='icon icon__lock icon--body'/>
+            );
+        }
+
+        return (
+            <div
+                onClick={this.handleClick}
+                className={className}
+                {...Suggestion.baseProps}
+            >
+                {icon}
+                {displayName}
+            </div>
+        );
+    }
+}
+
+let prefix = '';
+
+function channelSearchSorter(wrappedA, wrappedB) {
```

Comment:
`channelSearchSorter` can be pulled out from this component and be tested.  If not to be shared, you may add a new file like `/components/suggestion/suggestion_utils.js`

Reply:
I can move this function into a utils file, though I want to make sure it is written in a way that is general purpose enough to be used by the majority of `SuggestionProviders`. The use cases between the providers seem a bit varied, which is why I opted to keep this in the component file.

I based my implementation on the channel sorter in `switch_channel_provider`, and adjusted for this provider's requirements.

https://github.com/mattermost/mattermost-webapp/blob/5084127a59c0c48cec19c98979aa0bf833d6912d/components/suggestion/switch_channel_provider.jsx#L151

Reply:
Yup, I've noticed that too.  Thanks